### ISSUE TITLE
1.4.0 - Reload causes negative WaitGroup counter

### ISSUE THREAD

**Issue by inkychris** [coredns/coredns#2654]
Adding reload to a block in the `Corefile` causes the 1.4.0 server to crash when a change is detected and it attempts to reload.
```
# Corefile
. {
  hosts {
    10.0.0.1 git.example.co.uk
    10.0.0.2 ci.example.co.uk
    10.0.0.3 docker.example.co.uk
    fallthrough
  }
  forward . 8.8.8.8
  log
  errors
  reload
}
```
I'm running this with docker compose:
```
Starting dns-server_server_1 ... done
Attaching to dns-server_server_1
server_1  | .:53
server_1  | 2019-03-05T07:52:18.819Z [INFO] plugin/reload: Running configuration MD5 = 44e2f8a3e8c1a35f11d60b4fe7643f0f
server_1  | 2019-03-05T07:52:18.842Z [INFO] CoreDNS-1.4.0
server_1  | 2019-03-05T07:52:18.842Z [INFO] linux/amd64, go1.12, 8dcc7fc
server_1  | CoreDNS-1.4.0
server_1  | linux/amd64, go1.12, 8dcc7fc
server_1  | 2019-03-05T07:52:18.842Z [INFO] plugin/reload: Running configuration MD5 = 44e2f8a3e8c1a35f11d60b4fe7643f0f
server_1  | [INFO] Reloading
server_1  | [INFO] Reloading
server_1  | panic: sync: negative WaitGroup counter
server_1  |
server_1  | goroutine 60 [running]:
server_1  | sync.(*WaitGroup).Add(0xc000188450, 0xffffffffffffffff)
server_1  |     /tmp/tmp.We6hJEcGiI/go/src/sync/waitgroup.go:74 +0x135
server_1  | sync.(*WaitGroup).Done(...)
server_1  |     /tmp/tmp.We6hJEcGiI/go/src/sync/waitgroup.go:99
server_1  | github.com/coredns/coredns/core/dnsserver.(*Server).Stop.func1(0xc000188420, 0xc0004c0780)
server_1  |     /tmp/tmp.We6hJEcGiI/g/src/github.com/coredns/coredns/core/dnsserver/server.go:172 +0x40
server_1  | created by github.com/coredns/coredns/core/dnsserver.(*Server).Stop
server_1  |     /tmp/tmp.We6hJEcGiI/g/src/github.com/coredns/coredns/core/dnsserver/server.go:171 +0x7c
dns-server_server_1 exited with code 2
```
Just changing the image version in the compose file to 1.3.1 fixes the issue:
```
Status: Downloaded newer image for coredns/coredns:1.3.1
Recreating dns-server_server_1 ... done
Attaching to dns-server_server_1
server_1  | .:53
server_1  | 2019-03-05T08:03:03.235Z [INFO] CoreDNS-1.3.1
server_1  | 2019-03-05T08:03:03.235Z [INFO] linux/amd64, go1.11.4, 6b56a9c
server_1  | CoreDNS-1.3.1
server_1  | linux/amd64, go1.11.4, 6b56a9c
server_1  | 2019-03-05T08:03:03.235Z [INFO] plugin/reload: Running configuration MD5 = 86c4d93f41d9988a418b6110bda4c40c
server_1  | [INFO] Reloading
server_1  | 2019-03-05T08:03:37.684Z [INFO] plugin/reload: Running configuration MD5 = 443c6d0026e9a2249ecba7fe5b86e035
server_1  | [INFO] Reloading complete
server_1  | 2019-03-05T08:06:40.914Z [INFO] 10.11.0.160:57612 - 1489 "A IN docker.example.co.uk. udp 57 false 4096" NOERROR qr,aa,rd 66 0.000078451s
```

**Comment by fturib** [comment]
on the 1.4.0 output log your provide. it seems there is an extra event for reloading that happen before the complete start of CoreDNS

> server_1  | 2019-03-05T07:52:18.819Z [INFO] plugin/reload: Running configuration MD5 = 44e2f8a3e8c1a35f11d60b4fe7643f0f

And then there is a second reload action later with the same file (same MD5). 

I am wondering if there is a change in the process of Caddy events ...

**Comment by miekg** [comment]
We did upgrade Caddy. Should be simple to try out.

Would also like a test where we can catch these, which, as discussed
before, can't be a unit test because of the entire process reloading

**Comment by fturib** [comment]
I am working on it ...
I mean of the fix. For the test, I guess that needs to be added on the CI.

**Comment by fturib** [comment]
Exact ... Caddy is now generating the `InstanceStartupEvent` at start of the server.
We MUST not execute it in our main() any more...

**Comment by fturib** [comment]
Will send a PR.

**Comment by miekg** [comment]
fixed via #2656
